src/main/java/com/aws/greengrass/deployment/DeploymentService.java

```diff
@@ -628,13 +626,31 @@ private void createNewDeployment(Deployment deployment) {
             }
         }
 
-
-        Future<DeploymentResult> process = executorService.submit(deploymentTask);
+        // Submit deployment task and finish current deployment (persist metadata) when complete.
+        CompletableFuture<DeploymentResult> process =
+                deploymentStatusKeeper.submitDeploymentTask(executorService, deploymentTask)
+                        .exceptionally(exception -> {
+                            logger.atError().log("Error in Completing Deployment", exception);
+                            updateDeploymentResultAsFailed(deployment, deploymentTask, false, exception);
+                            return null;
+                        }).whenComplete((result, error) -> {
+                            if (error != null) {
+                                logger.atError().log("Error in Completing Deployment", error);
+                                updateDeploymentResultAsFailed(deployment, deploymentTask, false, error);
+                            } else {
+                                if (currentDeploymentTaskMetadata != null) {
+                                    try {
+                                        finishCurrentDeployment();
+                                    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)